src/centrality.rs

```diff
@@ -0,0 +1,307 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use crate::digraph;
+use crate::graph;
+
+use pyo3::prelude::*;
+use pyo3::Python;
+
+use hashbrown::HashMap;
+use petgraph::graph::NodeIndex;
+use petgraph::visit::{
+    GraphBase,
+    GraphProp, // allows is_directed
+    IntoNeighborsDirected,
+    IntoNodeIdentifiers,
+    NodeCount,
+    NodeIndexable,
+};
+
+// Correspondence to notation in Brandes 2001
+//
+// P -- predecessors
+// S -- sorted_by_distance,
+//      vertices in order of non-decreasing distance from s
+// Q -- Q
+// sigma -- sigma
+// delta -- delta
+// d -- distance
+
+pub fn betweenness_centrality<G>(
+    graph: G,
+    endpoints: bool,
+    normalized: bool,
+) -> Vec<Option<f64>>
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + NodeCount
+        + GraphProp
+        + GraphBase<NodeId = NodeIndex>,
+    // rustfmt deletes the following comments if placed inline above
+    // + IntoNodeIdentifiers // for node_identifiers()
+    // + IntoNeighborsDirected // for neighbors()
+    // + NodeCount // for node_count
+    // + GraphProp // for is_directed
+{
+    let mut max_index = 0;
+    for node_s in graph.node_identifiers() {
+        let is: usize = graph.to_index(node_s);
+        if is > max_index {
+            max_index = is;
+        }
+    }
+    max_index += 1;
```

Comment:
You can replace this with:

```suggestion
    let max_index = graph.node_bound();
```

The result of `node_bound()` is the current max index in the graph: https://docs.rs/petgraph/0.6.0/petgraph/stable_graph/struct.StableGraph.html#method.node_bound It does require the `NodeIndexable` trait but you already have that as a trait bound for `G`.

Reply:
I knew there must be something like this. I was grepping for things with "max" in it.